### ISSUE TITLE
PolygonZone: estimate how much of the zone is occupied

### ISSUE THREAD

**Issue by stevensagaert** [roboflow/supervision#1449]
### Search before asking

- [X] I have searched the Supervision [issues](https://github.com/roboflow/supervision/issues) and found no similar feature requests.

### Description

Next to the count of the objects in the zone, it would be useful to know how much of the zone area is occupied by the objects.

### Use case

e.g. for loading zones in warehouses

### Additional

Basically I've implemented it myself outside of the PolygonZone class. The core of what is needed is these two functions:

def polygon_to_binary_mask(img,contour):
    """convert bounding polygon coord to binary mask"""
    # Create binary mask
    b_mask = np.zeros(img.shape[:2], np.uint8)

    #  Extract contour result
    #contour = c.masks.xy.pop()
    #  Changing the type
    contour = contour.astype(np.int32) #Note: This must be int32 and not unit8!!!
    #  Reshaping
    contour = contour.reshape(-1, 1, 2)

    # Draw contour onto mask
    mask = cv2.drawContours(b_mask, [contour], -1, (1, 1, 1), cv2.FILLED)
    return mask

#strategy: 
#1) make union of binary masks of detected objects (only the classes for consideration in zone)
#2) make binary mask of the zone
#3) do and of zone mask & object masks
#4) sum the 1 pixels
def calculate_overlap_area(zone_mask, masks,show_plot=True):
    """calculate how much % of the zone is occupied"""
    #create one mask from the object masks as the union
    union_mask=np.bitwise_or.reduce(np.stack(masks),axis=0)
    #for debugging
    if show_plot:
        plt.title("union object mask")
        plt.imshow(union_mask,cmap='gray')
        plt.show()
    

    #do the bitwise and between union_mask & zone_mask
    overlap_mask=np.bitwise_and(union_mask,zone_mask)
    print(f"overlap_mask shape {overlap_mask.shape}")
    if show_plot:
        plt.title("overlap mask")
        plt.imshow(union_mask,cmap='gray')
        plt.show()
    
    overlap_size= np.sum(overlap_mask)
    zone_size=np.sum(zone_mask)
    return 100*overlap_size/zone_size

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by Sapienscoding** [comment]
Hi @stevensagaert, I'd like to help by submitting a PR

**Comment by LinasKo** [comment]
Hi @stevensagaert 👋 

Apologies for the delay. If you're still keen, we'd be happy to include area occupation stats in our PolygonZone.

**Comment by AyuK03** [comment]
Hi, it is my first time contributing to open source project and I would like to work on this issue. Please do guide me if possible. Thank you!

**Comment by stevensagaert** [comment]
Hi,
I'll make some time in 2 weeks or so. Right now I'm too busy at work.
________________________________

**Comment by AHB102** [comment]
@stevensagaert  Can I resolve this issue ?

**Comment by stevensagaert** [comment]
You mean you want to implement it or just close it? I haven't had time to do it yet.

**Comment by AHB102** [comment]
@stevensagaert I would like to try to implement it.

**Comment by stevensagaert** [comment]
Ok go ahead 🙂. The bulk of the functionality is in the 2 functions I put in this issue. The rest should be just glue code.
________________________________

**Comment by LinasKo** [comment]
Hi @AHB102 👋 

Assigning this to you.

Note that Detection can take one of 3 forms:
* segmentation - these will contain the `mask` field
* object detection - these will have the `xyxy` field, and not `mask`.
* oriented bounding box - these will have a `(N, 4, 2)`-shaped polygon in `detections.data[CLASS_NAME_DATA_FIELD]`.  (`CLASS_NAME_DATA_FIELD` == `"xyxyxyxy"` and resides in `config.py`)

Ideally, this method should work regardless of detection type, but it's okay if it returns a slightly-approximated result.

You may find the `polygon_to_mask` method helpful. it resides in: `supervision/detection/utils.py`.

**Comment by AHB102** [comment]
@LinasKo Thanks for the insight 😁, will get started on this.

**Comment by AHB102** [comment]
@LinasKo Hi 👋,I haven't written much code for this feature. I mostly used Steven's existing code, but I added checks for the detection cases based on your recommendations. Had some questions : 
1)Type hinting: I wrote this feature in vanilla Python . I'm new to type hinting, is it still okay to use it ?
2)Code location: I put the feature code in “supervision/detection/utils.py”, outside of the PolygonZone class. Is that the right place?
3)Testing: How do I test the feature? Can I use import statements, or is there a preferred way to write tests before submitting a pull request (PR)?

**Comment by LinasKo** [comment]
Hi @AHB102 👋

1. Ideally, we'd love a bit of type hinting at least in function definitions. For arguments, it's as simple as adding `: int`, `: str`, `: List[str]` and so on. For numpy arrays it's fine to say `np.ndarray`. Return type is ` -> int:` and so on. You can do ` -> None:` if it returns nothing. I'd encourage you to try!
2. I think so, but I'll need to see the PR first.
3. We encourage making a Google Colab and showing how it works. If you want, you can use this [Starter Template](https://colab.research.google.com/drive/1rin7WrS-UvVIe-_Gfxmu-yVslGphOq89#scrollTo=pjmCrNre2g58). There's no PolygonZone examples, but you can find some in the [docs](https://supervision.roboflow.com/latest/detection/tools/polygon_zone/), the [cheatsheet](https://roboflow.github.io/cheatsheet-supervision/) or [a notebook](https://github.com/roboflow/notebooks/blob/main/notebooks/how-to-detect-and-count-objects-in-polygon-zone.ipynb).
Try to come up with a situation where you can detect something in an image, and actually check the polygon zone occupancy. I suggest using Ultralytics for detection at the moment. Let me know how it goes, and if you can't figure it out 😉

**Comment by AHB102** [comment]
Okay nice! Will definitely make the type hinting work, and can I commit the changes and make a PR ?

**Comment by LinasKo** [comment]
Yes, absolutely. Typically contributors fork the repo, make the changes there on a branch such as `feat/estimate-polygonzone-occupation`, and then open a PR via github website to supervision's `develop` (there should be a button, as soon as you push to your own fork).

**Comment by AHB102** [comment]
@LinasKo I got a bit lost in the first PR. I believe this one is correct. Thanks for your patience😅

**Comment by LinasKo** [comment]
Thank you @AHB102!

I'll have a look as soon as I can, likely next week.

**Comment by SkalskiP** [comment]
@stevensagaert, @Sapienscoding, @AyuK03, @AHB102 I'm curious about what real-life use cases you are working on where information about zone occupancy level is interesting?

**Comment by stevensagaert** [comment]
loading zone in a warehouse
________________________________
